### PR TITLE
send installed name in manifest from targets

### PR DESCRIPTION
**Pull request by patriotyk**

### CODE REVIEW

**Comment by cajun-rat** [comment]
I'm not sure that we are heading for the right solution here, because we are still trying to parse the filename that is reported by the director:

~~~
+  std::size_t pos = ref_name.find_last_of("-");
+  if (pos == std::string::npos) throw std::runtime_error("Invalid refname");
~~~

I think the process should be:

1. Get the OSTree hash from `hashes.sha256`

~~~
"qemux86-64-ota-f2879dbbc2506c7e69fafc06541592d18d73b325edea6fa83950e76877f630f8" : 
                     ^^ We need to consider this as an opaque string, and must not parse it
		{
			"custom" : 
			{
				"ecuIdentifier" : "",
				"targetFormat" : ""
			},
			"hashes" : 
			{
				"sha256" : "F2879DBBC2506C7E69FAFC06541592D18D73B325EDEA6FA83950E76877F630F8"
                                 ^^^^ We should use this instead
			},
			"length" : 0
		},
~~~

2. Pass that to OSTree to install.
3. When we PUT the manifest to the server, we should:
4. Get the hash of the next boot from OSTree. It is important that we get this from OSTree and don't store it in `FSStorage`, because systems outside our control might rollback the deployment to a previous version
5. Look up that hash in the current images/targets.json
6. Find the filename (e.g. "qemux86-64-ota-f2879dbbc2506c7e69fafc06541592d18d73b325edea6fa83950e76877f630f8")
7. Report that filename directly to the server without trying to parse it

**Comment by patriotyk** [comment]
But the first part of filename is the branch_name (qemux86-64-ota in your example) and we need it for ostree install. So we still need to extract this part

File: .travis.yml [hunk 1]
```diff
@@ -5,14 +5,14 @@ branches:
...
+  - DOCKERFILE=Dockerfile SCRIPT=src/scripts/coverage.sh
+  - DOCKERFILE=Dockerfile.nop11 SCRIPT=src/scripts/test.sh
+  - DOCKERFILE=Dockerfile.deb-stable SCRIPT=src/scripts/test.sh DARGS="-eBUILD_ONLY=1"
+  - DOCKERFILE=Dockerfile.deb-unstable SCRIPT=src/scripts/test.sh DARGS="-eBUILD_ONLY=1"
 services:
   - docker
 script:
   - docker pull advancedtelematic/aktualizr:$CACHE
-  - docker build --cache-from advancedtelematic/aktualizr:$CACHE -t advancedtelematic/aktualizr -f ${DOCKERFILE} .
+  - docker build -t advancedtelematic/aktualizr -f ${DOCKERFILE} .
```
Comment (cajun-rat) [inline .travis.yml]: Why are these changes needed?  Is there something broken with the caching behaviour?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -80,9 +79,12 @@ if(BUILD_P11)
     endif(TEST_PKCS11_MODULE_PATH)
 endif(BUILD_P11)
 # set the name of the executable
-add_executable(aktualizr main.cc)
+if(BUILD_OSTREE)
+ add_executable(aktualizr main.cc ostree.cc)
+else(BUILD_OSTREE)
+ add_executable(aktualizr main.cc)
+endif(BUILD_OSTREE)
```
Comment (patrickvacek) [inline src/CMakeLists.txt]: This is super-trivial, but the indenting is inconsistent. Should be four spaces. The vim magic on the last line echoes that.

**Comment by patrickvacek** [comment]
So I'm seeing some strange things when I try to test this. For example:

`Oct 10 09:21:10 qemux86-64 aktualizr[266]: [2017-10-10 09:21:09.612778]: Error of pulling image: Invalid character '67' in rev 'C32054FDBA572CE4DE4139050E5D4728F0D9D83DD1DD9F398DB17A106A3EA4D5'`

Also, many of these:

`Oct 10 09:21:09 qemux86-64 aktualizr[266]: libostree HTTP error from remote aktualizr-remote for <https://e433a4df-1000-4683-b6d9-55450c60124f.tcpgw.prod01.advancedtelematic.com/treehub/summary.sig>: Server returned HTTP 404`

I'm testing against our production server. Have you seen something like this?

**Comment by patrickvacek** [comment]
Related note, I also cannot actually update my device. The UI just shows my updates as failing but it does not give a reason.

File: src/ostree.cc [hunk 3]
```diff
@@ -88,15 +88,15 @@ bool Ostree::addRemote(OstreeRepo *repo, const std::string &remote, const std::s
...
-      ref_name(ref_name_in),
-      branch_name(branch_name_in),
-      refhash(refhash_in),
-      description(desc_in),
-      pull_uri(treehub_in) {}
+OstreePackage::OstreePackage(const std::string &ref_name_in, const std::string &refhash_in, const std::string &desc_in,
+                             const std::string &treehub_in)
+    : ref_name(ref_name_in), refhash(refhash_in), description(desc_in), pull_uri(treehub_in) {
+  std::size_t pos = ref_name.find_last_of("-");
+  if (pos == std::string::npos) throw std::runtime_error("Invalid refname");
```
Comment (cajun-rat) [inline src/ostree.cc]: I'd like to get rid of this requirement that the name in targets.json has a specific form--but maybe this can wait for a future PR

**Comment by patrickvacek** [comment]
Now seeing this:
```
Oct 12 09:30:59 qemux86-64 aktualizr[300]: [2017-10-12 09:30:57.670597]: Error of pulling image: Invalid character '65' in rev '48A88D23B9B411D8CC9054C36768CA8FAB3D2C61DF69A0EA803C37B0E5340452'
```
Are capital letters in the rev a problem or something?

**Comment by patriotyk** [comment]
There is some strange  issue, running make qa gives me no differences in git diff. So I cannot update PR to fix build. 
@patrickvacek could You try to reproduce this issue again? I have updated this PR and now it should work.

**Comment by patrickvacek** [comment]
I think I fixed the formatting problem. Not sure why that happens occasionally; I still haven't figured out why our versions produce different output.

Anyway, now when I try to push an update, I see this in the aktualizr log: `Oct 13 07:30:26 qemux86-64 aktualizr[306]: [2017-10-13 07:30:26.709071]: Error of pulling image: Commit 269e5f62ad9f3afd7213694beb84ce3a2428490ef86c1b10d640db663e6b65ac: commit has no requested ref ‘aktualizr’ in ref binding metadata (‘build2_pyro2’)`

**Comment by patriotyk** [comment]
@patrickvacek In the email You wrote that You was able to install update, but with some issues.

**Comment by patrickvacek** [comment]
@patriotyk Ah, sorry for the confusion. The update install I had mentioned in the email was not using your branch. I was just excited that the QA server was working at all. I have not managed to get a successful update with your branch on any server.

**Comment by patrickvacek** [comment]
Still getting basically the same error: `Oct 19 09:02:12 qemux86-64 aktualizr[298]: aa57da: commit has no requested ref ‘aktualizr’ in ref binding metadata (‘build2_pyro2’)`. Interestingly, the QA server is now giving me a similar message:
```Result code 4
Commit df57718c1d321d1f6c4fff6ee39a1265905b9e44ba702631d3fd0dc405aa57da: commit has no requested ref ‘aktualizr’ in ref binding metadata (‘build2_pyro2’)```

@cajun-rat @OYTIS Could one of you try this and see if it works for you?

**Comment by patriotyk** [comment]
I have built a fresh image and everything works well for me(there was one issue but not related to this problem). Could You explain detailed steps how You doing that? How do You change bitbake recipe, how upload second packages for update on it.

**Comment by patrickvacek** [comment]
1. I have a checkout of the pyro branch of updater-repo.
2. I put `BRANCH_pn-aktualizr = "issue/PRO-3758/put_names"` and `require classes/sota_bleeding.inc` in my `local.conf`.
3. I run `bitbake core-image-minimal` as normal.
4. I run `../meta-updater/scripts/run-qemu-ota core-image-minimal` as normal.
5. I log in to the device once it boots and run `journalctl -fu aktualizr`.
6. I make a small change to `IMAGE_INSTALL_append` in `local.conf`, usually adding or removing `man`.
7. I again run `bitbake core-image-minimal` as normal.
8. I find the device in ATS Garage, select the most recently built package, and click install.
9. I watch the output of `journalctl` on the device and look for errors.
10. I refresh ATS Garage until I see an error or success.

**Comment by cajun-rat** [comment]
I get the same error as @patrickvacek 

```
Commit a63efc942a7d4f3c2e930f22ec4a2c787a1bd72f172c65c0fc0dbad035e8fa38: commit has no requested ref ‘aktualizr’ in ref binding metadata (‘qemux86-64’)
```

I'm using https://app.atsgarage.com

**Comment by cajun-rat** [comment]
@patriotyk What version of aktualizr is getting built into the image?  If `aktualizr --version` doesn't match the commit you expect, then maybe you need `require classes/sota_bleeding.inc` and `BRANCH_pn-aktualizr = "issue/PRO-3758/put_names"`  in your local.conf, to force building the latest commit on your branch

**Comment by cajun-rat** [comment]
@patriotyk What version of aktualizr is getting built into the image?  If `aktualizr --version` doesn't match the commit you expect, then maybe you need `require classes/sota_bleeding.inc` and `BRANCH_pn-aktualizr = "issue/PRO-3758/put_names"`  in your local.conf, to force building the latest commit on your branch

**Comment by patriotyk** [comment]
I verified this version and it was correct (hash was from my lastest commit in brach)

**Comment by patriotyk** [comment]
Now it works

**Comment by patrickvacek** [comment]
Now I'm seeing this (on the QA server):
```
Error of pulling image: Commit c7c541654fe8456d0a7bc63ce8f6e7075497505f3a0a0c3ffdff14fe8c5a68c0: commit has no requested ref ‘qemux86-64’ in ref binding metadata (‘build2_pyro2’)
```

**Comment by patriotyk** [comment]
Have You checked the version hash? It should be 8854f82

**Comment by patrickvacek** [comment]
I see it as [
8df30bb ](https://github.com/advancedtelematic/aktualizr/pull/361/commits/8df30bb38623df1038b46ff6596bc3e341606c48). It looks like the commit you mentioned is one behind the commit I'm looking at (the one that was originally my clang-format fixup).

**Comment by patriotyk** [comment]
Sorry, Yes I wrote the wrong commit hash. But it is very strange. Do You run aktualizr manually from command-line?

**Comment by patrickvacek** [comment]
No, I just let it run automatically from the service.

**Comment by patriotyk** [comment]
Ok, because I found the similar issue, when run two instances of aktualizr, but You run only one instance

**Comment by patrickvacek** [comment]
Also, I just realized this breaks `uptane_ci_test` (because `getEcu` is no longer part of `OstreePackage`). That test is only ran if `SOTA_PACKED_CREDENTIALS` is provided to CMake, which I just so happened to have set now but not before.

**Comment by patrickvacek** [comment]
In response to my own question after consulting with Anton: yes, it does matter that my package name is not the same as the hardware ID. When I fix that (`OSTREE_BRANCHNAME = "qemux86-64"`), it works fine. I think we can stop worrying about that for now.

The last thing preventing merging, though, is still the `uptane_ci_test` issue. @patriotyk Can you take a look at that?

**Comment by patriotyk** [comment]
ok, done
